master-server/src/main/java/com/replicated_log/master_server/service/impl/BrokerServiceImpl.java

```diff
@@ -0,0 +1,132 @@
+        addNotSentItemsToBrokerMap(secondaryServerAddressAdded, itemsFromBeginning);
+        if (!itemsFromBeginning.isEmpty()) {
+
+            //when secondary address is added to Master, we start the broker(if it is not started yet).
+            startBroker();
+        }
+        return secondaryServerAddressAdded;
+    }
+
+    @Override
+    public synchronized void addItemToBrokerMap(Item item) {
+        LOG.info("S--> BrokerServiceImpl: Adding item " + item);
+        brokerRepository.addItemToBrokerMap(item);
+
+        //when first item is added to Master, we start the broker.
+        startBroker();
+    }
+
+    @Override
+    public void removeItemFromBrokerMap(Item item, Address secondaryServerAddress) {
+        LOG.info("S--> BrokerServiceImpl: Removing item " + item + " for Address " + secondaryServerAddress);
+        brokerRepository.removeItemFromBrokerMap(item, secondaryServerAddress);
+    }
+
+    @Override
+    public synchronized Set<Item> addNotSentItemsToBrokerMap(Address secondaryServerAddress, Set<Item> notSentItems) {
+        LOG.info("S--> BrokerServiceImpl: Adding items " + notSentItems + " for Address " + secondaryServerAddress);
+        return brokerRepository.addNotSentItemsToBrokerMap(secondaryServerAddress, notSentItems);
+    }
+
+    @Override
+    public synchronized void publishToSecondaries() {
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (runBroker) {
+                    LOG.info("S--> BrokerServiceImpl: Publishing to Secondaries");
+                    runBroker = false;
+
+                    Map<Address, Set<Item>> brokerStorageMap = brokerRepository.getBrokerStorage();
+                    for (Map.Entry<Address, Set<Item>> entry : brokerStorageMap.entrySet()) {
+
+                        Address secondaryServerAddress = entry.getKey();
+                        Set<Item> items = entry.getValue();
+
+                        if (!items.isEmpty()) {
+                            runBroker = true;
+
+                            for (Item item : items) {
+                                if (!HealthStatus.UNHEALTHY.equals(secondaryServerAddress.getHealthStatus())) {
+                                    LOG.info("S--> BrokerServiceImpl: Publishing of item " + item + " to " + secondaryServerAddress + " (" + secondaryServerAddress.getHealthStatus() + ")");
+                                    publishItemToSecondary(item, secondaryServerAddress, 300);
+                                }
+                            }
+                        }
+                    }
+
+                    sleepMethod(1000);
+                }
+            }
+        }).start();
+    }
+
+    @Override
+    public void publishItemToSecondary(Item item, Address secondaryServerAddress, int sleepMillis) {
+        String baseUrl = secondaryServerAddress.getAddress();
+
+        MasterClient masterClient = new MasterClient(WebClient.create(baseUrl));
```

Comment:
try to avoid situation when one class with logic try to create new instance of other, you deprive of the opportunity to dynamic injection/changing implementation of some logic. Either use configuration for it, or factory class. This is not responsibility of this class

Reply:
Created MasterClient bean in configuration  class and then used it.

Reply:
Good point 👍